subprojects/core-api/src/main/java/org/gradle/api/file/SourceDirectorySetAdapter.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file;
+
+import org.gradle.api.provider.Provider;
+import org.gradle.internal.instrumentation.api.annotations.BytecodeUpgrade;
+
+class SourceDirectorySetAdapter {
+    @BytecodeUpgrade
+    static Provider<Directory> getClassesDirectory(SourceDirectorySet sourceDirectorySet) {
```

Comment:
We also want to do something similar with [SourceSetOutput#getClassesDirs](https://github.com/gradle/gradle/blob/cb0c615fb8e3690971bb7f89ad80f58943360624/platforms/jvm/platform-jvm/src/main/java/org/gradle/api/tasks/SourceSetOutput.java#L95), updating it to return a ConfigurableFileCollection instead of a FileCollection. Would we create a separate adapter class for that?